File: backend/src/main/java/com/gxs/bank/service/CardService.java
package com.gxs.bank.service;

import com.gxs.bank.dto.request.CardApplyRequest;
import com.gxs.bank.exception.BadRequestException;
import com.gxs.bank.exception.ResourceNotFoundException;
import com.gxs.bank.model.Card;
import com.gxs.bank.model.User;
import com.gxs.bank.repository.CardRepository;
import com.gxs.bank.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;
import java.util.Random;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class CardService {

    private final CardRepository cardRepository;
    private final UserRepository userRepository;

    public Card applyForCard(UUID userId, CardApplyRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        if (user.getKycStatus() != com.gxs.bank.model.User.KycStatus.VERIFIED) {
            throw new BadRequestException("KYC must be verified before applying for a card");
        }

        Card.CardType cardType;
        try {
            cardType = Card.CardType.valueOf(request.getCardType().toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new BadRequestException("Invalid card type. Must be one of: DEBIT, CREDIT, PREPAID, FLEXI");
        }

        BigDecimal creditLimit = switch (cardType) {
            case CREDIT -> new BigDecimal("200000.00");
            case FLEXI -> new BigDecimal("5000.00");
            case PREPAID -> new BigDecimal("50000.00");
            default -> BigDecimal.ZERO;
        };

        // Auto-assign card network
        Card.CardNetwork network = (cardType == Card.CardType.CREDIT || cardType == Card.CardType.FLEXI)
                ? Card.CardNetwork.MASTERCARD : Card.CardNetwork.VISA;

        Card card = Card.builder()
                .user(user)
                .cardNumberLast4(String.format("%04d", new Random().nextInt(10000)))
                .cardHolderName(user.getFullName().toUpperCase())
                .cardType(cardType)
                .cardNetwork(network)
                .status(Card.Status.PENDING)
                .creditLimit(creditLimit)
                .currentBalance(BigDecimal.ZERO)
                .cashbackEarned(BigDecimal.ZERO)
                .build();

        return cardRepository.save(card);
    }

    public List<Card> getUserCards(UUID userId) {
        return cardRepository.findByUserId(userId);
    }

    public Card getCard(UUID cardId, UUID userId) {
        Card card = cardRepository.findById(cardId)
                .orElseThrow(() -> new ResourceNotFoundException("Card not found"));
        if (!card.getUser().getId().equals(userId)) {
            throw new BadRequestException("Card does not belong to user");
        }
        return card;
    }

    public Card toggleFreeze(UUID cardId, UUID userId) {
        Card card = getCard(cardId, userId);
        if (card.getStatus() == Card.Status.ACTIVE) {
            card.setStatus(Card.Status.FROZEN);
        } else if (card.getStatus() == Card.Status.FROZEN) {
            card.setStatus(Card.Status.ACTIVE);
        } else {
            throw new BadRequestException("Cannot toggle freeze on a cancelled card");
        }
        return cardRepository.save(card);
    }

    public Card updateSettings(UUID cardId, UUID userId, java.util.Map<String, Object> settings) {
        Card card = getCard(cardId, userId);
        if (settings.containsKey("isInternationalEnabled")) {
            card.setIsInternationalEnabled(Boolean.parseBoolean(settings.get("isInternationalEnabled").toString()));
        }
        if (settings.containsKey("isOnlineEnabled")) {
            card.setIsOnlineEnabled(Boolean.parseBoolean(settings.get("isOnlineEnabled").toString()));
        }
        if (settings.containsKey("isContactlessEnabled")) {
            card.setIsContactlessEnabled(Boolean.parseBoolean(settings.get("isContactlessEnabled").toString()));
        }
        return cardRepository.save(card);
    }
}

File: backend-python/com/gxs/bank/service/CardService.py
from decimal import Decimal
import random

from com.gxs.bank.dto.request.CardApplyRequest import CardApplyRequest
from com.gxs.bank.exception.BadRequestException import BadRequestException
from com.gxs.bank.exception.ResourceNotFoundException import ResourceNotFoundException
from com.gxs.bank.model.Card import Card
from com.gxs.bank.model.User import User
from com.gxs.bank.repository.CardRepository import CardRepository
from com.gxs.bank.repository.UserRepository import UserRepository


class CardService:
    def __init__(self, db):
        self.db = db
        self.cardRepository = CardRepository(db)
        self.userRepository = UserRepository(db)

    def applyForCard(self, userId: str, request: CardApplyRequest):
        user = self.userRepository.findById(userId)
        if user is None:
            raise ResourceNotFoundException("User not found")

        if user.kycStatus != User.KycStatus.VERIFIED:
            raise BadRequestException("KYC must be verified before applying for a card")

        try:
            cardType = Card.CardType(request.cardType.upper())
        except ValueError:
            raise BadRequestException("Invalid card type. Must be one of: DEBIT, CREDIT, PREPAID, FLEXI") from None

        creditLimit = {
            Card.CardType.CREDIT: Decimal("200000.00"),
            Card.CardType.FLEXI: Decimal("5000.00"),
            Card.CardType.PREPAID: Decimal("50000.00"),
        }.get(cardType, Decimal("0.00"))

        network = (
            Card.CardNetwork.MASTERCARD
            if cardType in {Card.CardType.CREDIT, Card.CardType.FLEXI}
            else Card.CardNetwork.VISA
        )

        card = Card(
            user=user,
            userId=user.id,
            cardNumberLast4=f"{random.randint(0, 9999):04d}",
            cardHolderName=user.fullName.upper(),
            cardType=cardType,
            cardNetwork=network,
            status=Card.Status.PENDING,
            creditLimit=creditLimit,
            currentBalance=Decimal("0.00"),
            cashbackEarned=Decimal("0.00"),
        )
        self.cardRepository.save(card)
        self.db.commit()
        self.db.refresh(card)
        return card

    def getUserCards(self, userId: str):
        return self.cardRepository.findByUserId(userId)

    def getCard(self, cardId: str, userId: str):
        card = self.cardRepository.findById(cardId)
        if card is None:
            raise ResourceNotFoundException("Card not found")
        if card.userId != userId:
            raise BadRequestException("Card does not belong to user")
        return card

    def toggleFreeze(self, cardId: str, userId: str):
        card = self.getCard(cardId, userId)
        if card.status == Card.Status.ACTIVE:
            card.status = Card.Status.FROZEN
        elif card.status == Card.Status.FROZEN:
            card.status = Card.Status.ACTIVE
        else:
            raise BadRequestException("Cannot toggle freeze on a cancelled card")
        self.db.commit()
        self.db.refresh(card)
        return card

    def updateSettings(self, cardId: str, userId: str, settings: dict):
        def _parse_bool(raw_value):
            if isinstance(raw_value, bool):
                return raw_value
            return str(raw_value).strip().lower() == "true"

        card = self.getCard(cardId, userId)
        if "isInternationalEnabled" in settings:
            card.isInternationalEnabled = _parse_bool(settings["isInternationalEnabled"])
        if "isOnlineEnabled" in settings:
            card.isOnlineEnabled = _parse_bool(settings["isOnlineEnabled"])
        if "isContactlessEnabled" in settings:
            card.isContactlessEnabled = _parse_bool(settings["isContactlessEnabled"])
        self.db.commit()
        self.db.refresh(card)
        return card
